src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs

```diff
@@ -1785,36 +1785,72 @@ private TypeSymbol GetAccessedType(AccessExpressionSyntax syntax, IDiagnosticWri
                 : baseType;
         }
 
-        private static TypeSymbol GetNamedPropertyType(PropertyAccessSyntax syntax, TypeSymbol baseType, IDiagnosticWriter diagnostics) => UnwrapType(baseType) switch
+        private static TypeSymbol? TryGetReadablePropertyType(ObjectType objectType, string propertyName)
         {
-            ErrorType error => error,
-            TypeSymbol withErrors when withErrors.GetDiagnostics().Any() => ErrorType.Create(withErrors.GetDiagnostics()),
-
-            TypeSymbol original when TypeHelper.TryRemoveNullability(original) is TypeSymbol nonNullable => EmitNullablePropertyAccessDiagnosticAndEraseNullability(syntax, original, nonNullable, diagnostics),
-
-            // the property is not valid
-            // there's already a parse error for it, so we don't need to add a type error as well
-            ObjectType when !syntax.PropertyName.IsValid => ErrorType.Empty(),
+            if (objectType.Properties.TryGetValue(propertyName) is {} property && !property.Flags.HasFlag(TypePropertyFlags.WriteOnly))
+            {
+                return property.TypeReference.Type;
+            }
 
-            ObjectType objectType => TypeHelper.GetNamedPropertyType(objectType,
-                syntax.PropertyName,
-                syntax.PropertyName.IdentifierName,
-                syntax.IsSafeAccess || TypeValidator.ShouldWarnForPropertyMismatch(objectType),
-                diagnostics),
+            if (objectType.AdditionalPropertiesType is {} additionalPropertiesType && !objectType.AdditionalPropertiesFlags.HasFlag(TypePropertyFlags.WriteOnly))
+            {
+                return additionalPropertiesType.Type;
+            }
 
-            UnionType unionType when unionType.Members.All(x => x is ObjectType)
-                => TypeHelper.CreateTypeUnion(unionType.Members.Select(member => GetNamedPropertyType(syntax, member.Type, diagnostics))),
+            // TODO could consider returning the null type here, but need to be careful about downstream impact.
+            // See https://github.com/Azure/bicep/issues/14059 for example
```

Comment:
@jeskew what are your thoughts on the risk here?

Reply:
I think we do want to use an explicit `null`, since we have linters that will propose a fix from `<base>.<property>` to `<base>.?<property>` if the property has a type like `string | null`. But this should only occur if `<base>` is an object. So the type of `foo.bar` should be the following (mapping `typeof(foo)` => `typeof(foo.bar)`):

* `{ bar: string } | { bar: int }` => `string | int`
* `{ bar: string } | {}` => `string | null`
* `{ otherProp: string } | { otherProp: int }` => `error`
* `{ bar: string } | string` => `error`

Like you say in the comment, though, we can use `any` now and refine it later if you want to get this out immediately.

Reply:
Thanks! I think I'd like to leave it more permissive for now, because making this change breaks the 2 samples given in the issue report.

In particular, this becomes quite annoying to handle:
```bicep
var items = [
  { obj: 1 }
  { obj: 2 }
  { obj: 3 }
  { obj: 4 }
  { obj: 5, x: 5 }
]
var s = sort(
  filter(
    map(items, x => x.obj < 2 ? { order: 1, value: x } : x.obj > 4 ? { order: 2, value: x } : {}),
    x => !(empty(x))
  ),
  (arg1, arg2) => arg1.order < arg2.order
)
```

Because the impact on types of `!(empty(x))` isn't understood by the `filter` function, you end up having to do a lot of work in the `sort` function - e.g. I think you'd either need `(arg1, arg2) => any(arg1).order < any(arg2).order` or `(arg1, arg2) => (arg1.?order ?? 0) < (arg2.?order ?? 0)`